### PR TITLE
[curl] Update

### LINKED ISSUE #14030: Unable to compile curl with LDAPS
Host Environment:
Windows 10 ver 1909, compiler : VS 2019

I tried to compiled Curl with vcpkg , and was unable to get the protocol LDAPS to be enabled in curl (I tried many options). The configuration seems to enable LDAP but not LDAPS. 

Steps to reproduce : 
Compile curl with tool (and openssl but this does not seem to matter)(in x64 Native commands tools): 
vcpkg install curl[tool,openssl] --triplet x64-windows

Then goes to package packages\curl_x64-windows\tools\curl and do:
curl ldaps://localhost
curl ldap://localhost
Result :
curl: (1) Protocol "ldaps" not supported or disabled in libcurl
curl: (7) Failed to connect to localhost port 389: Connection refused

I expect the first message to be somewhat the same as the second (here LDAPS is not activated).

When I compile LibCurl with nmake (going in the winbuild directory and doing something like :
"nmake /f Makefile.vc mode=dll WITH_SSL=dll SSL_PATH=[YOUR_SSL_PATH] VC=16 MACHINE=x64 GEN_PDB=yes ENABLE_UNICODE=yes") I am able to get the LDAPS support. 

At this point I am not sure if this is vcpkg related or just a bad makefile from LibCurl, but as nmake in winbuild works I guess that this may be related to vcpkg. 

config-x64-windows-out.log parts that can be relevant : 
-- Looking for cldap_open in wldap32;winmm;ws2_32;advapi32;crypt32;OpenSSL::SSL;OpenSSL::Crypto
-- Looking for cldap_open in wldap32;winmm;ws2_32;advapi32;crypt32;OpenSSL::SSL;OpenSSL::Crypto - found
-- Looking for include file winldap.h
-- Looking for include file winldap.h - not found
-- Looking for include file winber.h
-- Looking for include file winber.h - not found
-- Looking for include file ldap_ssl.h
-- Looking for include file ldap_ssl.h - not found
-- Looking for include file ldapssl.h
-- Looking for include file ldapssl.h - not found
[...]
-- Enabled protocols: DICT FILE FTP FTPS GOPHER HTTP HTTPS IMAP IMAPS LDAP POP3 POP3S RTSP SMB SMBS SMTP SMTPS TELNET TFTP
(no LDAPS here, it indicates it does not find winldap but I think it findsit because LDAP is activated at the end)

### PR DESCRIPTION
**Pull request by dg0yt**
**Describe the pull request**

- #### What does your PR fix?  
  - [x] Modernize and update port file
  - [x] Improve pc file (no implict libs, no duplication, correctly choose optimized vs. debug e.g. for zlib, flatten transitive dependencies)
  - [x] Install rectified `curl-config` for release and debug
  - [x] Update curl (related: #16126)
  - [x] Add WinIDN support
  - [x] Revise libidn2 support (related: #19002)
  - [x] Fix winssl dependencies (#19359)
  - [x] Make WinLDAP an explicit, non-default feature
         Even with WinLDAP, there is no LDAPS support (#14030), which is now documented in the feature description.
         There are no vcpkg ports for supporting LDAP support on UWP and non-Windows platforms. With the explicit winldap feature, we can reliably deactivate ldap dependency lookups on these platforms (#17919).

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

### CODE REVIEW

**Comment by dg0yt** [comment]
Curl now also supports "WinIDN" so this can be used for Windows instead of libidn2.
I will try to make IDN support work for mingw, too. 
Libidn2 needs a review of how it publishes its icon and unistring dependencies.

**Comment by dg0yt** [comment]
I won't care much with the osx gdal issue before #17698 is merged.

**Comment by dg0yt** [comment]
Waiting for merge of
#19002 for libidn2 support
#17698 for osx gdal issue

**Comment by dg0yt** [comment]
I'm looking into upgrading to curl 0.78.0, and I want to get rid of the nghttp2 patch by #19163.

File: ports/curl/0020-fix-pc-file.patch [hunk 1]
```diff
@@ -1,16 +1,59 @@
...
 +    if(CMAKE_MATCH_1 IN_LIST CMAKE_C_IMPLICIT_LINK_FRAMEWORK_DIRECTORIES)
 +      set(_lib "-framework ${CMAKE_MATCH_2}")
 +    else()
 +      set(_lib "-framework ${_lib}")
 +    endif()
-   endif()
++  endif()
    if(_lib MATCHES ".*/.*" OR _lib MATCHES "^-")
      set(LIBCURL_LIBS          "${LIBCURL_LIBS} ${_lib}")
+   else()
```
Comment (NancyLi1013) [inline ports/curl/0020-fix-pc-file.patch:59]: Is it possible to submit these changes to upsream?

Comment (dg0yt) [reply ports/curl/0020-fix-pc-file.patch:59]: Yes, I am willing to do that. However, this shall not delay the merge of the PR.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
Have you tested these features in this PR? @dg0yt

**Comment by dg0yt** [comment]
> Have you tested these features in this PR?

Regarding `*idn*`:
I tested building the features on mingw (static and dynamic) and on linux, and I reviewed the pc files.
I'm sure I also had some rebuilds of gdal while the curl idn features were installed.
I did not run real tests.

**Comment by NancyLi1013** [comment]
All features `curl[brotli,c-ares,core,idn2,mbedtls,non-http,openssl,schannel,ssh,ssl,sspi,tool,winidn,winssl] `have passed with the following triplets:

- x86-windows
- x64-windows-static

For feature `sectransp`, it only supports osx platform
`sectransp is not supported on non-Apple platforms`

For feature `http2`, it build successfully on x86-windows and failed on x64-windows-static with the following errors:

```
2019: unresolved external symbol __imp_nghttp2_session_set_local_window_size referenced in function Curl_http2_switched
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_is_fatal referenced in function on_frame_recv
src\curl.exe : fatal error LNK1120: 34 unresolved extelibcurl-d.lib(version.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_version referenced in function curl_version_info
libcurl-d.lib(http2.c.obj) : error LNK2001: unresolved external symbol __imp_nghttp2_version
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_new referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_del referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_send_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_on_frame_recv_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_on_data_chunk_recv_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_on_stream_close_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_on_begin_headers_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_on_header_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_callbacks_set_error_callback referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_client_new referenced in function http2_init
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_del referenced in function http2_stream_free
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_send referenced in function Curl_http2_done
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_mem_recv referenced in function h2_process_pending_input
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_resume_data referenced in function Curl_http2_done_sending
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_want_read referenced in function should_close_session
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_want_write referenced in function Curl_http2_done_sending
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_get_stream_user_data referenced in function on_frame_recv
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_set_stream_user_data referenced in function Curl_http2_switched
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_get_stream_local_window_size referenced in function Curl_http2_stream_pause
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_get_remote_settings referenced in function on_frame_recv
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_upgrade2 referenced in function Curl_http2_switched
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_pack_settings_payload referenced in function Curl_http2_request_upgrade
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_strerror referenced in function Curl_http2_switched
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_http2_strerror referenced in function http2_handle_stream_close
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_priority_spec_init referenced in function h2_pri_spec
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_submit_request referenced in function http2_send
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_submit_priority referenced in function h2_session_send
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_submit_rst_stream referenced in function Curl_http2_done
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_submit_settings referenced in function Curl_http2_switched
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_submit_ping referenced in function get_transfer
libcurl-d.lib(http2.c.obj) : error LNK2019: unresolved external symbol __imp_nghttp2_session_check_request_allowed referenced in function h2_process_pending_input
libcurl-d.lib(http2.c.obj) : error LNKrnals
ninja: build stopped: subcommand failed.
```

Could you please take a look? @dg0yt

**Comment by dg0yt** [comment]
> failed on x64-windows-static with the following errors

@NancyLi1013 Can you please verify that you are you using the latest version of port nghttp2? The static build should `define NGHTTP2_STATICLIB` at the end of `include/nghttp2/nghttp2ver.h`.

**Comment by dg0yt** [comment]
> > failed on x64-windows-static with the following errors
> 
> @NancyLi1013 Can you please verify that you are you using the latest version of port nghttp2? The static build should `define NGHTTP2_STATICLIB` at the end of `include/nghttp2/nghttp2ver.h`.

In CI, `curl[core,http2,non-http,openssl,schannel,ssl,sspi,winssl]:x64-windows-static-md` built just fine:  
https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/57226/logs/117 
(The agent for x64-windows-static died too early, but it reached curlpp so I assume there is no issue with `curl[http2]`.)

**Comment by dg0yt** [comment]
Reset to draft for the following changes:
- fix winssl dependencies, #19359.
- review `non-http` feature, cf. #14030 and #17919.

What is wrong `non-http`:
- It is a default feature, making it hard to opt out. (Every port which depends on default features of curl pulls in non-http.) 
  This wouldn't be a problem but
- It enables auto-discovery of libraries, so it may pick up system libraries, e.g. LDAP.

To ensure proper binary packages, we must disable each feature which has additional dependencies unless 
- it is enabled by a non-default feature, and
- that feature adds proper dependencies

An additional feature "system-libraries" may allow the discovery of non-vcpkg libraries if desired. (Prototype: port gdal.)

**Comment by dg0yt** [comment]
>     * review `non-http` feature

I guess I checked this before: I don't see what other system libraries would be picked up by curl ATM. There are many default-OFF switches so this should all be safe.

**Comment by NancyLi1013** [comment]
Have you fixed the problem on feature `http2`?

**Comment by dg0yt** [comment]
> Have you fixed the problem on feature `http2`?

I asked you to double-check that you tested the feature with the latest version of port nghttp2. May I kindly repeat my request?
CI didn't find any problem when I did a test build with the http2 enabled, as I reported in another comment here.

**Comment by NancyLi1013** [comment]
> > Have you fixed the problem on feature `http2`?
> 
> I asked you to double-check that you tested the feature with the latest version of port nghttp2. May I kindly repeat my request?
> CI didn't find any problem when I did a test build with the http2 enabled, as I reported in another comment here.

I tested this feature based on your branch. `nghttp2 `is 1.42.0, the latest version in vcpkg master branch is 1.44.0. I think you need to merge master to your branch.

**Comment by dg0yt** [comment]
> I think you need to merge master to your branch.

Oh, you test the HEAD of this branch without master. That's different from CI, of course. I did the merge now. But you could do this as well when testing. (Or should do it that way? It you test just the branch, it *is different* from what the user gets after the PR is merged.)

**Comment by NancyLi1013** [comment]
> > I think you need to merge master to your branch.
> 
> Oh, you test the HEAD of this branch without master. That's different from CI, of course. I did the merge now. But you could do this as well when testing. (Or should do it that way? It you test just the branch, it _is different_ from what the user gets after the PR is merged.)

I have update to the latest changes to test the feature `http2`. It can be built successfully on x64-windows-static now.

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
LGTM now, thanks for your updates and improvements @dg0yt.

**Comment by JackBoosY** [comment]
Fixes #14030.

**Comment by BillyONeal** [comment]
Thanks for your hard work!

**Pull request merged**